src/vector_of_vectors.jl

```diff
@@ -0,0 +1,96 @@
+# Data structure that behaves like a `Vector{Vector}`, but uses a contiguous memory layout.
+# Similar to `VectorOfVectors` of ArraysOfArrays.jl, but allows to resize the inner vectors.
+struct DynamicVectorOfVectors{T, ARRAY2D, ARRAY1D} <: AbstractVector{Array{T, 1}}
+    backend::ARRAY2D # Array{T, 2}, where each column represents a vector
+    length_::Base.RefValue{Int32} # Number of vectors
+    lengths::ARRAY1D # Array{Int32, 1} storing the lengths of the vectors
+end
+
+function DynamicVectorOfVectors{T}(; max_outer_length, max_inner_length) where {T}
+    backend = Array{T, 2}(undef, max_inner_length, max_outer_length)
+    length_ = Ref(zero(Int32))
+    lengths = zeros(Int32, max_outer_length)
+
+    return DynamicVectorOfVectors{T, typeof(backend), typeof(lengths)}(backend, length_,
+                                                                       lengths)
+end
+
+@inline Base.size(vov::DynamicVectorOfVectors) = (vov.length_[],)
+
+@inline function Base.getindex(vov::DynamicVectorOfVectors, i)
+    (; backend, lengths) = vov
+
+    @boundscheck checkbounds(vov, i)
+
+    return view(backend, 1:lengths[i], i)
+end
+
+@inline function Base.push!(vov::DynamicVectorOfVectors, vector::AbstractVector)
+    (; backend, length_, lengths) = vov
+
+    # This data structure only supports one-based indexing
+    Base.require_one_based_indexing(vector)
+
+    # Activate a new column of `backend`
+    j = length_[] += 1
+    lengths[j] = length(vector)
+
+    # Fill the new column
+    for i in eachindex(vector)
+        backend[i, j] = vector[i]
+    end
+
+    return vov
+end
+
+@inline function Base.push!(vov::DynamicVectorOfVectors, vector::AbstractVector, vectors...)
+    push!(vov, vector)
+    push!(vov, vectors...)
+end
+
+# `push!(vov[i], value)`
+@inline function pushat!(vov::DynamicVectorOfVectors, i, value)
```

Comment:
We could use `push!(vov[i], value)` and `deleteat!(vov[i], j)` by returning a wrapper around a view, which also stores the `vov` and the column where the view comes from. This wrapper would behave like a regular view, but allow `push!` and `deleteat!`. Not sure if that would be more elegant than this here. We only use this data structure internally, anyway.

Reply:
> We only use this data structure internally, anyway.

Famous last words.